rfcs/2021-12-20-10517-llvm-backend-for-vrl.md

```diff
@@ -0,0 +1,560 @@
+
+The result pointer can be temporarily changed using `context.set_result_ref()`. This mechanism allows the parent expression to call `emit_llvm` on the child while controlling where the machine code emitted for the child expression stores the result. This is useful, e.g. when emitting a binary operation where both of its operands need to be computed first.
+
+Calling `context.context_ref()` returns a reference to the VRL [`Context`](https://github.com/vectordotdev/vector/blob/f1404bea186ba83c4426a32bbef3f633c17cf4d2/lib/vrl/compiler/src/context.rs#L5-L9) that is provided by the `remap` transform as a function argument.
+
+For anything less trivial than emitting branches or calling functions, we want to leverage the Rust compiler. For one, this allows us to not concern ourselves with memory layout and doesn't force us to define an FFI when we only use basic integer types or pointers/references. It also provides us with Rust's memory safety guarantees for large parts of the emitted LLVM IR.
+
+Most expressions which rely predominantly on precompiled Rust build LLVM IR similar to the following
+
+```rust
+let fn_ident = "vrl_resolved_initialize";
+let fn_impl = ctx
+    .module()
+    .get_function(fn_ident)
+    .ok_or(format!(r#"failed to get "{}" function"#, fn_ident))?;
+ctx.builder()
+    .build_call(fn_impl, &[result_ref.into()], fn_ident);
+```
+
+which will emit the LLVM instruction
+
+```llvm
+call void @vrl_resolved_initialize(%"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>"* %result)
+```
+
+This call refers to a function compiled by Rust in the same LLVM module. However, we don't need to actually call this function, the implementation can be inlined and optimized by LLVM since the whole source code is known. This way of emitting LLVM IR is merely very convenient to stitch together code fragments.
+
+For temporary values needed e.g. in binary operations we can allocate uninitialized stack values:
+
+```rust
+let result_temp_ref = ctx.build_alloca_resolved("temp")?;
+```
+
+which will emit the LLVM instruction
+
+```llvm
+%temp = alloca %"std::result::Result<vrl_compiler::Value, vrl_compiler::ExpressionError>", align 8
+```
+
+it is our responsibility to initialize and drop the value accordingly. This can be accomplished by calling the implementations for `vrl_resolved_initialize` and `vrl_resolved_drop` shown further below.
+
+Constants can be moved into the LLVM module by consuming the constant value of type `T`, transmuting it to `[i8]` and transmuting it back to `T` when unloading the LLVM module. This is safe since Rust's semantics allow all types to be moved in memory unless they are `Pin`. Writing constants into the LLVM module has the benefit of allowing LLVM to apply constant folding at compile time.
```

Comment:
By placing the constant inside the LLVM module, LLVM can do deeper analysis and potentially simplify or even completely remove instructions and branches. If we don't move the constant's memory into the LLVM module and instead reference to constants that we allocated in the calling function, how is LLVM going to know its contents?

For some of the optimizer passes see:

- `-instcombine`: [Combine redundant instructions](https://llvm.org/docs/Passes.html#instcombine-combine-redundant-instructions)
- `-reassociate`: [Reassociate expressions](https://llvm.org/docs/Passes.html#reassociate-reassociate-expressions)
- `-sccp`: [Sparse Conditional Constant Propagation](https://llvm.org/docs/Passes.html#passes-sccp)

Stuff like figuring that you can bitshift instead of multiplying, e.g. `x << 3` instead of `x * 8` or that `if (x * 2 + 1) % 2 == 0` always evaluates to false, so we always take the `else` branch and don't need to evaluate the expression. If, from the point of view of LLVM, it were `x * y` there was no way for it to find out that it's a power of two.

Another reference to look at, the "Trivial Constant Folding" chapter in the Kaleidoscope tutorial: https://llvm.org/docs/tutorial/MyFirstLanguageFrontend/LangImpl04.html#id2.

Reply:
Yeah, I understand the concept of constant folding, and needing to be able to give LLVM a value as constant, rather than a named variable, etc, for it to be able to do constant folding... but I guess what I'm confused about is that you specifically mention going from `T` to `[i8]` and back to `T`.

If the value can be constant, that would say to me that it's a constant on the Rust side: a bitwise-copyable scalar (`u32`) or a type that can be const constructed, and thus was also `Copy`, etc.  Which then forms the basis of my question: do we actually need to be doing any sort of memory movement back _out_ of LLVM?  Like I guess I'm mostly just confused around if you're saying we would allocate values in Rust and then essentially make Rust forget about them and "transfer" them to LLVM in some way?

Reply:
Oh, I see, I picked up the question on the wrong end 😄

Good that you followed up with more questions, I hope I can see what's unclear now.

> that would say to me that it's a constant on the Rust side: a bitwise-copyable scalar (`u32`) or a type that can be const constructed, and thus was also `Copy`, etc.

That's the cool thing: Rust's memory model allows moving values in memory, even if they are not `const`, `Clone` or any of that. They merely need not to be `Pin`. By writing the bytes that make up the constant values to the LLVM module, we are effectively doing a move in memory.

> do we actually need to be doing any sort of memory movement back out of LLVM?

Yes, we do. I guess I was too vague with "back to `T` when unloading the LLVM module", I'll make that section more specific.

> Like I guess I'm mostly just confused around if you're saying we would allocate values in Rust and then essentially make Rust forget about them and "transfer" them to LLVM in some way?

I'm not sure if that's a question. But yes, we do exactly that:

- create constant value in Rust
- transmute value to `[i8]` in Rust
- forget `[i8]` in Rust
- write those `[i8]` it to LLVM module
- LLVM might be able to see through the value and do some constant folding
- dynamically load LLVM module which makes that location live in memory again
- when unloading the dynamic library, "remember" the `[i8]` in Rust by copying it to memory and transmuting it back to `T`
- drop `T` to call its destructors

Reply:
Cleared that up in 42ed20b92e731aee7f9487afca85334e7545a7fc.